### DISCUSSION TITLE
Is `Tags.WAVELENGTH` still necessary?

### DISCUSSION THREAD

**Discussion by TomTomRixRix** [IMSY-DKFZ/simpa#327] [Ideas]
Currently there are two different tags: `Tags.WAVELENGTHS` and `Tags.WAVELENGTH`. While the former is used to define all wavelengths used for simulation or processing, the latter is used to specify the currently processed wavelength or the wavelength to be visualized or tested. `Tags.WAVELENGTH` is iteratively updated in the global settings in the `simulate` method while looping through all wavelengths.

:bulb: **Idea:**
Remove the `Tags.WAVELENGTH` and:
1. In `examples` and `tests` replace the `Tags.WAVELENGTH` with the last entry from `Tags.WAVELENGTHS` as this value will anyways be set during simulation in `simulate()`.
2. In all `simulation modules` and `processing components` we could pass the currently used wavelength from the `simulate` method to each adapter. Currently everything happens in the `run` method of the adapter, however, this should maybe be independent of wavelengths. The wavelength is currently only used for reading and writing the required data for each adapter. Therefore, two new abstract methods could be introduced: `load_data` and `save_data`. We could either <br>
  a) write something like this

   ```python
   data = pipeline_element.load_data(wavelength)
   data = pipeline_element.run(digital_device_twin, data)
   pipeline_element.save_data(wavelength, data)
   ```
   where we return the loaded and processed data and give it to the next function via arguments, or 

   b) use such a solution
   ```python
   pipeline_element.load_data(wavelength)
   pipeline_element.run(digital_device_twin)
   pipeline_element.save_data(wavelength)
   ```
   where the data is stored in attributes of the adapter.

**Pros and Cons**
* (b) would be cleaner as the data doesn't leave the adapter
* (a) would provide the opportunity to run the adapter without loading the data with the `load_data` method of the adapter but providing your own data to the adapter, e.g. 
  ```python
   my_adapter = Adapter()
   my_adapter.run(my_digital_device_twin, my_data)
   ``

**Things to consider**
* How is the RAM usage affected be these options?

**Comment by leoyala** [comment]
I like the idea of removing redundancy in the code. I would just make the methods more explicit about what they do, such as `load_adapter_data` or so. I am more of a fan of saving the data as attributes, which would imply also creating a new method to set it, for example `set_adapter_data` in case the data wants to be passed to the pipeline manually.

**Comment by jgroehl** [comment]
Yes, sounds logical. Have you already looked into this?

**Reply by TomTomRixRix** [reply]
No, so far this was only an idea which I've put here to keep it somewhere.
But removing `Tags.WAVELENGTH` would also solve one of the issues (read-only access to settings) of #29
